### PR TITLE
Better HTTP responses and simpler schema generation

### PR DESCRIPTION
**Pull request by marcelkornblum**
> This is a follow-on from #39 and #38  (and git branched from `consistent-service-interfaces` which is in turn branched off `simpler-services` used in those respective PRs - please merge both of those before this.

The HTTP response codes we're generating are not tight enough; we were allowing for example someone to POST a user ID that already existed on the system; although we were taking no action we were returning a 200. I've updated the codes to be more accurate.

In addition our schema approach is very boilerplate driven and clunky; it would be a lot neater and easier to follow if we used Django-ninja's inbuilt ModelSchema. This is an attempt to simplify that side of things, as well as putting the schemas closer to what they represent.

### CODE REVIEW

File: scim/api.py [hunk 1]
```diff
@@ -1,27 +1,43 @@
...
-
-response_codes = frozenset({200, 201})
+    try:
+        return profile_services.combined.get_by_id(id)
+    except Profile.DoesNotExist:
+        return 404, {"message": "No user found with that ID"}
 
 
-@router.post("scim/v2/Users", response={response_codes: SCIMUserOut})
+@router.post("scim/v2/Users", response={201: ProfileMinimal, 409: Error})
```
[the remark this answers is not in the document]
Comment (marcelkornblum) [reply scim/api.py:31]: 1. Do we need to return anything for our current use cases? If not why spend time implementing it?
2. If so, is there a reason not to use the profile?
I'm happy to use something else if needed but I suspect it's not actually needed, and I'd like a justification for the complication of the the current setup with all the boilerplate and dicts and things

[the remark this answers is not in the document]
Comment (marcelkornblum) [reply scim/api.py:31]: I also think we will have a separate profile endpoint - doesnt mean we cant necessarily return it through scim as well (though TBH I havent got my head around scim too much)

File: scim/api.py [hunk 2]
```diff
@@ -28,15 +29,23 @@ def get_user(request, id: str):
...
     try:
         user = core_services.create_user(
             scim_user.externalId,
             PROFILE_TYPE_STAFF_SSO,
-            first_name=scim_user.name,
-            last_name=scim_user.name,  # @TODO do we get the names split out?
-            emails=scim_user.emails,
+            first_name=scim_user.name.givenName,
+            last_name=scim_user.name.familyName,
+            emails=emails,
```
Comment (CamLamb) [inline scim/api.py]: We lose the "primary" flag from SSO when we convert emails to a list of strings, something to think about

File: core/tests/test_core_service.py [hunk 3]
```diff
@@ -11,22 +11,22 @@ class TestCoreService(TestCase):
 
     @pytest.mark.django_db
     @pytest.mark.skip()
-    def test_create_user(self):
+    def test_create_identity(self):
         # User is created
-        user = core_services.create_user(
+        user, created = core_services.create_identity(
```
Comment (marcelkornblum) [inline core/tests/test_core_service.py:16]:
```suggestion
        user = core_services.create_identity(
```

File: core/services.py [hunk 4]
```diff
@@ -22,11 +22,13 @@ def create_user(
...
+
     user = user_services.create(sso_email_id=id)
     profile_services.create_from_sso(
         id,
         first_name,
         last_name,
         emails,
     )
+
     return user
```
Comment (marcelkornblum) [inline core/services.py:34]: return combined profile

File: profiles/models/generic.py [hunk 5]
```diff
@@ -0,0 +1,31 @@
...
+
+
+class Email(AbstractHistoricalModel):
+    address = models.EmailField(validators=[EmailValidator()], unique=True)
+
+    def __str__(self):
+        return self.address
+
+
+class Profile(AbstractHistoricalModel):
```
Comment (marcelkornblum) [inline profiles/models/generic.py:20]: move to combined.py

File: profiles/models/abstract.py [hunk 6]
```diff
@@ -0,0 +1,23 @@
...
+
+class AbstractHistoricalModel(models.Model):
+    class Meta:
+        abstract = True
+
+    history = HistoricalRecords(
+        inherit=True,
+    )
+
+
```
Comment (marcelkornblum) [inline profiles/models/abstract.py:15]:
```suggestion
# @TODO discuss why not inherit from AbstractHistoricalModel
```

File: scim/api.py [hunk 7]
```diff
@@ -1,29 +1,60 @@
...
 
 
 router = Router()
 
 
-@router.get("scim/v2/Users/{id}", response=SCIMUserOut)
+@router.get(
+    "scim/v2/Users/{id}",
+    response={
+        200: ProfileMinimal,
```
Comment (marcelkornblum) [inline scim/api.py:19]: use correct schema `GetUserResponse`

File: scim/api.py [hunk 8]
```diff
@@ -1,29 +1,60 @@
...
+        404: Error,
+    },
+)
+# @TODO check if we actually need this method, if not let's drop it or move out of SCIM
```
Comment (marcelkornblum) [inline scim/api.py:23]:
```suggestion
```

File: scim/schemas.py [hunk 9]
```diff
@@ -1,18 +1,18 @@
 from dataclasses import dataclass
```
Comment (marcelkornblum) [inline scim/schemas.py:1]: add a comment about the commenting-out unused but protocol-compliant fields

File: scim/schemas.py [hunk 10]
```diff
@@ -22,105 +22,74 @@ class Email:
...
+    # x509Certificates
+
+
+class CreateUserRequest(ScimUserSchema): ...
 
 
-class SCIMUserOut(Schema):
-    schemas: List = ["urn:ietf:params:scim:schemas:core:2.0:User"]
-    id: str = Field(alias="sso_email_id")
+class CreateUserResponse(ScimUserSchemaRequired):
```
Comment (marcelkornblum) [inline scim/schemas.py:75]: add another layer in here for MinimalUserResponse witht he mappings in

File: scim/schemas.py [hunk 11]
```diff
@@ -22,105 +22,74 @@ class Email:
...
-        if type(obj) is get_user_model():
-            return [PhoneNumber("0123456789", "work", True)]
-        else:
-            if "phoneNumbers" in obj:
-                return obj["phoneNumbers"]
-            else:
-                return None
+        combined_profile = combined.get_by_id(obj.sso_email_id)
+        return Name(
+            givenName=combined_profile.first_name,
```
Comment (marcelkornblum) [inline scim/schemas.py:87]: update these to work with Profile

File: scim/api.py (hunk 2 above)
Comment (marcelkornblum) [reply scim/api.py:56]: resolving this as relevant to Alsen's upcoming PR...

**Review by alsyx — APPROVED** [review]

**Pull request merged**